09-maze_solver/progress/internal/solver/imagefile.go

```diff
@@ -0,0 +1,95 @@
+package solver
+
+import (
+	"fmt"
+	"image"
+	"image/gif"
+	"image/png"
+	"log/slog"
+	"os"
+	"strings"
+)
+
+// openMaze returns a RGBA png image.
+func openMaze(inputPath string) (*image.RGBA, error) {
+	_, err := os.Stat(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to check input file %s: %w", inputPath, err)
+	}
+
+	fd, err := os.Open(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open input image at %s: %w", inputPath, err)
+	}
+	defer fd.Close()
+
+	img, err := png.Decode(fd)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load input image from %s: %w", inputPath, err)
+	}
+
+	// Using RGBAAt() instead of At() saves a lot of time.
+	rgbaImage, ok := img.(*image.RGBA)
+	if !ok {
+		return nil, fmt.Errorf("this isn't a RGBA image")
+	}
+
+	return rgbaImage, nil
+}
+
+// SaveSolution saves the image as a PNG file with the solution path highlighted.
+func (s *Solver) SaveSolution(outputPath string) error {
+	_, err := os.Stat(outputPath)
```

Comment:
This check is unnecessary, and potentially even wrong. The file could be created just before `os.Create` and hence this check is logically racy. Instead check the error from `os.Create` for special cases, if they are relevant for the problem.